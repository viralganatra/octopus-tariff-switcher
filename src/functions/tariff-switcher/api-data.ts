import { formatISO } from 'date-fns';
import { UnknownProductError } from '../../errors/unknown-product-error';
import { UnknownTariffError } from '../../errors/unknown-tariff-error';
import { getMsFromApiIsoString, roundTo4Digits } from '../../utils/helpers';
import { logger } from '../../utils/logger';
import {
  fetchAccountInfo,
  fetchAllProducts,
  fetchSmartMeterTelemetry,
  fetchTodaysStandingCharge,
  fetchTodaysUnitRatesByTariff,
  fetchToken,
} from './queries';
import type { TariffSelector } from '../../types/tariff';

const [TARIFF_AGILE, TARIFF_COSY] = ['Agile Octopus', 'Cosy Octopus'] as const;
const [TARIFF_CODE_AGILE, TARIFF_CODE_COSY] = ['AGILE-', 'COSY-'] as const;

type Tariff = typeof TARIFF_AGILE | typeof TARIFF_COSY;

let token: string;

export function getOppositeTariff(currentTariff: Tariff) {
  return currentTariff === TARIFF_AGILE ? TARIFF_COSY : TARIFF_AGILE;
}

export async function getToken() {
  if (token) {
    return token;
  }

  logger.info('Getting token from API');

  const data = await fetchToken();

  token = data.obtainKrakenToken.token;

  return token;
}

export async function getAccountInfo() {
  const token = await getToken();
  const results = await fetchAccountInfo({ token });

  logger.info('Got account info from API', { apiResponse: results });

  const [electricityAgreement] = results.account.electricityAgreements;
  const { tariffCode, standingCharge } = electricityAgreement.tariff;
  const [{ deviceId }] = electricityAgreement.meterPoint.meters[0].smartDevices;
  // tariffCode is in the format E-1R-COSY-22-12-08-A and should always be a non empty string,
  // but typescript casts this to at to string | undefined
  const regionCode = tariffCode.at(-1) as string;
  const normalisedStandingCharge = roundTo4Digits(standingCharge);

  let currentTariff: Tariff;

  if (tariffCode.includes(TARIFF_CODE_AGILE)) {
    currentTariff = TARIFF_AGILE;
  } else if (tariffCode.includes(TARIFF_CODE_COSY)) {
    currentTariff = TARIFF_COSY;
  } else {
    throw new UnknownTariffError(`Current tariff is neither Agile nor Cosy, it is: ${tariffCode}`);
  }

  return {
    currentTariff,
    regionCode,
    deviceId,
    currentStandingCharge: normalisedStandingCharge,
  };
}

export async function getTodaysConsumptionInHalfHourlyRates({
  deviceId,
}: {
  deviceId: string;
}) {
  const token = await getToken();

  const today = formatISO(new Date(), { representation: 'date' });
  const startDate = `${today}T00:30:00Z`;
  const endDate = `${today}T23:59:59Z`;

  const { smartMeterTelemetry } = await fetchSmartMeterTelemetry({
    token,
    startDate,
    endDate,
    deviceId,
  });

  logger.info('Got half hourly consumption data from API', {
    apiResponse: smartMeterTelemetry,
  });

  const data = smartMeterTelemetry.map(({ costDeltaWithTax, ...halfHourlyUnitRate }) => ({
    ...halfHourlyUnitRate,
    unitCostInPence: costDeltaWithTax,
    readAtMs: getMsFromApiIsoString(halfHourlyUnitRate.readAt),
  }));

  return data;
}

export async function getTodaysUnitRatesByTariff({ tariffCode, productCode }: TariffSelector) {
  const results = await fetchTodaysUnitRatesByTariff({ tariffCode, productCode });

  logger.info(
    `Getting todays unit rates for tariff code: ${tariffCode} and product code: ${productCode}`,
    {
      apiResponse: results[0],
    },
  );

  const unitRatesWithMs = results.map(({ value_inc_vat, ...halfHourlyUnitRate }) => ({
    ...halfHourlyUnitRate,
    validFromMs: getMsFromApiIsoString(halfHourlyUnitRate.valid_from),
    validToMs: getMsFromApiIsoString(halfHourlyUnitRate.valid_to),
    unitCostInPence: value_inc_vat,
  }));

  return unitRatesWithMs;
}

async function getTodaysStandingCharge({ tariffCode, productCode }: TariffSelector) {
  const results = await fetchTodaysStandingCharge({ tariffCode, productCode });

  logger.info(
    `Got today's standing charge for tariff code: ${tariffCode} and product code: ${productCode} from API`,
    {
      apiResponse: results,
    },
  );

  const todaysStandingCharge = roundTo4Digits(results[0].value_inc_vat);

  return todaysStandingCharge;
}

export async function getPotentialRatesAndStandingChargeByTariff({
  regionCode,
  tariff,
}: {
  regionCode: string;
  tariff: Tariff;
}) {
  logger.info(`Getting todays rates for tariff: ${tariff} in region: ${regionCode}`);

  const allProducts = await fetchAllProducts();

  // Find the Octopus product for the tariff we're looking for
  const currentProduct = allProducts.find((product) => {
    return product.display_name === tariff && product.direction === 'IMPORT';
  });

  logger.info(`Found matching product based on ${tariff}`, {
    data: currentProduct,
  });

  // There should always be a product for Agile or Cosy
  if (!currentProduct) {
    throw new UnknownProductError(`Unable to find valid product using: ${tariff}`);
  }

  const tariffCode = currentProduct.code;
  // Residential tariffs are always E-1R
  const productCode = `E-1R-${tariffCode}-${regionCode}`;

  const [potentialUnitRates, potentialStandingCharge] = await Promise.all([
    getTodaysUnitRatesByTariff({ tariffCode, productCode }),
    getTodaysStandingCharge({ tariffCode, productCode }),
  ]);

  return { potentialUnitRates, potentialStandingCharge };
}
