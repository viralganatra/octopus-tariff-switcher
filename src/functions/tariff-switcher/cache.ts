import type { AllProducts } from './schema';

let token: string;
let allProducts: AllProducts = [];

export function getCachedToken() {
  return token;
}

export function setCachedToken(newToken: string) {
  token = newToken;
  return token;
}

export function getCachedProducts() {
  return allProducts;
}

export function setCachedProducts(products: AllProducts) {
  allProducts = products;
  return allProducts;
}
