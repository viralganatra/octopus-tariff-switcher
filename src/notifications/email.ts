import { z } from 'zod';
import { Resource } from 'sst';
import { API_MJML } from '../constants/api';
import { sendData } from '../utils/fetch';
import {
  getAlreadyCheapestTariffTemplate,
  getCheaperTariffTemplate,
  notWorthSwitchingTariffTemplate,
} from '../utils/email-template';
import { sendSparkPostEmail } from './sparkpost';
import { logger } from '../utils/logger';
import type { TariffContextWithCost } from '../types/tariff';
import type { EmailType } from '../types/email';

type SendEmail = {
  allTariffsByCost: TariffContextWithCost[];
  currentTariffWithCost: TariffContextWithCost;
  emailType: EmailType;
};

async function getHtml(content: string) {
  const mjmlAuth = `${Resource.MjmlAppId.value}:${Resource.MjmlSecretKey.value}`;

  logger.info(`API: Getting email template data via ${API_MJML}`);

  const schema = z.object({
    html: z.string(),
  });

  const results = await sendData({
    url: API_MJML,
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
      Authorization: `Basic ${btoa(mjmlAuth)}`,
    },
    body: { mjml: content },
  });

  const { html } = schema.parse(results);

  logger.info('API Response: Received email template data');

  return html;
}

export async function sendEmail({ allTariffsByCost, currentTariffWithCost, emailType }: SendEmail) {
  let emailSubject: string;
  let emailContent: string;

  switch (emailType) {
    case 'CHEAPER_TARIFF_EXISTS':
      emailSubject = 'Cheaper Tariff Found';
      emailContent = getCheaperTariffTemplate({
        allTariffsByCost,
        currentTariffWithCost,
        title: emailSubject,
      });
      break;
    case 'ALREADY_ON_CHEAPEST_TARIFF':
      emailSubject = 'You are already on the cheapest tariff';
      emailContent = getAlreadyCheapestTariffTemplate({
        allTariffsByCost,
        currentTariffWithCost,
        title: emailSubject,
      });
      break;
    case 'NOT_WORTH_SWITCHING_TARIFF':
      emailSubject = 'Not worth switching tariff';
      emailContent = notWorthSwitchingTariffTemplate({
        allTariffsByCost,
        currentTariffWithCost,
        title: emailSubject,
      });
      break;
  }

  const html = await getHtml(emailContent);

  return sendSparkPostEmail({ html, subject: emailSubject });
}
