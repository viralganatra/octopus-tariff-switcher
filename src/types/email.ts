import type { TariffContextWithCost } from './tariff';

export type EmailType =
  | 'CHEAPER_TARIFF_EXISTS'
  | 'ALREADY_ON_CHEAPEST_TARIFF'
  | 'NOT_WORTH_SWITCHING_TARIFF';

export type SendEmail = {
  allTariffsByCost: TariffContextWithCost[];
  currentTariffWithCost: TariffContextWithCost;
  emailType: EmailType;
};
