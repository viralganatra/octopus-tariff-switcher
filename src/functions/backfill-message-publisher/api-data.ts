import { isWithinInterval, parseISO } from 'date-fns';
import type { IsoDate } from '../../types/misc';
import type { ConsumptionIntervals, EletricityAgreements } from './schema';
import { chunkArray, processBatches } from '../../utils/batch';
import { TARIFFS, type TariffDisplayName } from '../../constants/tariff';
import { retryWithExponentialBackoff } from '../../utils/fetch';
import { fetchConsumption, fetchStandingCharge } from './queries';
import { fetchUnitRatesByTariff } from '../tariff-switcher/queries';
import type { TariffUnitRates } from '../tariff-switcher/schema';

export type TariffData = {
  isoDate: IsoDate;
  tariffCode: string;
  tariffName: TariffDisplayName;
  productCode: string;
  standingCharge: number;
  unitRates: TariffUnitRates;
  consumption: ConsumptionIntervals;
};

type ItemCacheBuilder = Pick<TariffData, 'isoDate' | 'tariffCode' | 'tariffName' | 'productCode'> &
  Partial<Omit<TariffData, 'isoDate' | 'tariffCode' | 'tariffName' | 'productCode'>>;

const BATCH_SIZE = 25;

export function findMatchingTariffForDate({
  pastTariffs,
  isoDate,
}: {
  pastTariffs: EletricityAgreements;
  isoDate: IsoDate;
}) {
  const targetDate = parseISO(isoDate);

  // Normalize target to start of day UTC
  const targetDateUtc = new Date(
    Date.UTC(targetDate.getUTCFullYear(), targetDate.getUTCMonth(), targetDate.getUTCDate()),
  );

  const matchingTariff = pastTariffs.find(({ validFrom, validTo }) => {
    const from = parseISO(validFrom);
    const to = parseISO(validTo);

    // isWithinInterval treats end as inclusive so we need to make it exclusive
    return (
      isWithinInterval(targetDateUtc, {
        start: from,
        end: to,
      }) && targetDateUtc.getTime() !== to.getTime()
    );
  });

  if (!matchingTariff) {
    throw new Error(`Unable to find matching tariff for date: ${isoDate}`);
  }

  return matchingTariff;
}

// Find the tariff code is in the format E-1R-COSY-22-12-08-A, extract COSY-22-12-08
// This isn't ideal, but there is no API endpoint that allows us to get a product code
// from a tariff code. Using the products endpoint with an available_at param doesn't work
// as you can be on an older tariff for a given date
function findProductCodeByTariffCode(tariffCode: string) {
  const regex = /(?:E-\dR-)(.*?)(?:-[A-Z])?$/;

  const parts = regex.exec(tariffCode);

  if (!parts || !parts[1]) {
    throw new Error(`No product code found from tariff code: ${tariffCode}`);
  }

  return parts[1];
}

export async function enrichDatesWithTariffData({
  dates,
  pastTariffs,
  mpan,
  serialNumber,
}: { dates: IsoDate[]; pastTariffs: EletricityAgreements; mpan: string; serialNumber: string }) {
  const itemCacheBuilder = new Map<IsoDate, ItemCacheBuilder>();
  const batchesOfDates = chunkArray(dates, BATCH_SIZE);

  // Build initial cache directly
  for (const isoDate of dates) {
    const { tariffCode } = findMatchingTariffForDate({ pastTariffs, isoDate });
    const productCode = findProductCodeByTariffCode(tariffCode);

    const tariff = TARIFFS.find(({ tariffCodeMatcher }) => tariffCode.includes(tariffCodeMatcher));

    if (!tariff) {
      throw new Error(`No matching tariff for: ${tariffCode}`);
    }

    itemCacheBuilder.set(isoDate, {
      isoDate,
      tariffCode,
      productCode,
      tariffName: tariff.displayName,
    });
  }

  // Fetch standing charge, unit rates & consumption in parallel batches
  const unitAndStandingChargesForDates = await processBatches({
    batches: batchesOfDates,
    processBatchItem: (isoDate) =>
      retryWithExponentialBackoff(async () => {
        const item = itemCacheBuilder.get(isoDate);

        if (!item || !item.productCode) {
          throw new Error(`Missing item or productCode for date: ${isoDate}`);
        }

        const { tariffCode, productCode } = item;

        const [standingCharge, unitRates, consumption] = await Promise.all([
          fetchStandingCharge({ isoDate, tariffCode, productCode }),
          fetchUnitRatesByTariff({ isoDate, tariffCode, productCode }),
          fetchConsumption({ isoDate, mpan, serialNumber }),
        ]);

        return { isoDate, standingCharge, unitRates, consumption };
      }),
  });

  // Populate item cache with standing charge, unit rates & consumption
  const itemCache = new Map<IsoDate, TariffData>();

  for (const {
    isoDate,
    standingCharge,
    unitRates,
    consumption,
  } of unitAndStandingChargesForDates) {
    const item = itemCacheBuilder.get(isoDate);

    if (!item || !item.productCode) {
      throw new Error(`Missing cache item for date: ${isoDate}`);
    }

    itemCache.set(isoDate, {
      isoDate,
      standingCharge,
      unitRates,
      consumption,
      tariffCode: item.tariffCode,
      tariffName: item.tariffName,
      productCode: item.productCode,
    });
  }

  return itemCache;
}
