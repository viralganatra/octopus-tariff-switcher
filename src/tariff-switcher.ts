import type { APIGatewayProxyEvent, APIGatewayProxyResult, Context } from 'aws-lambda';
import {
  acceptNewAgreement,
  getAccountInfo,
  getEnrollmentId,
  getPotentialRatesAndStandingChargeByTariff,
  getConsumptionInHalfHourlyRates,
  verifyNewAgreement,
} from './functions/tariff-switcher/api-data';
import {
  getDailyUsageCostByTariff,
  getDailyCostInPence,
} from './functions/tariff-switcher/cost-calculator';
import { penceToPoundWithCurrency, sleep } from './utils/helpers';
import { logger } from './utils/logger';
import { formatResponse } from './utils/format-response';
import { TARIFFS } from './constants/tariff';
import { sendEmail } from './notifications/email';
import { AgreementVerificationError } from './errors/agreement-verification-error';
import type { SendEmail } from './types/email';

function isDryRun() {
  return process.env.DRY_RUN === 'true';
}

function logSuccessAndRespond(successMessage: string) {
  const message = isDryRun() ? `DRY RUN: ${successMessage}` : successMessage;

  logger.info(message);

  return formatResponse(200, { message });
}

function sendTariffEmail(params: SendEmail) {
  return isDryRun() ? Promise.resolve() : sendEmail(params);
}

export async function tariffSwitcher(
  event: APIGatewayProxyEvent,
  context: Context,
): Promise<APIGatewayProxyResult> {
  logger.addContext(context);

  try {
    const { deviceId, currentStandingCharge, regionCode, currentTariff, productCode, mpan } =
      await getAccountInfo();

    const todaysConsumptionUnitRates = await getConsumptionInHalfHourlyRates({ deviceId });

    logger.info(`Recieved today's consumption unit rates`, {
      data: todaysConsumptionUnitRates,
    });

    const todayCostInPence = getDailyCostInPence({
      unitRates: todaysConsumptionUnitRates,
      standingCharge: currentStandingCharge,
    });

    const todaysCostInPounds = penceToPoundWithCurrency(todayCostInPence);

    logger.info(`Today's consumption cost is ${todaysCostInPounds}`);

    const currentTariffWithCost = {
      ...currentTariff,
      productCode,
      costInPence: todayCostInPence,
    };

    const comparedTariffs = [currentTariffWithCost];

    for (const tariff of TARIFFS) {
      if (tariff.id === currentTariff.id) {
        continue;
      }

      const { potentialUnitRates, potentialStandingCharge, potentialProductCode } =
        await getPotentialRatesAndStandingChargeByTariff({
          regionCode,
          tariff: tariff.displayName,
        });

      const potentialCostInPence = getDailyUsageCostByTariff({
        consumptionUnitRates: todaysConsumptionUnitRates,
        standingCharge: potentialStandingCharge,
        tariffUnitRates: potentialUnitRates,
      });

      comparedTariffs.push({
        ...tariff,
        costInPence: potentialCostInPence,
        productCode: potentialProductCode,
      });
    }

    // Find cheapest tariff
    const sortedTariffsByCost = comparedTariffs.toSorted((a, b) => a.costInPence - b.costInPence);

    logger.info('All tariffs by cost', {
      data: comparedTariffs,
    });

    const cheapestTariff = sortedTariffsByCost.at(0) ?? currentTariffWithCost;
    const cheapestTariffCostInPounds = penceToPoundWithCurrency(cheapestTariff.costInPence);

    if (cheapestTariff.id === currentTariff.id) {
      await sendTariffEmail({
        allTariffsByCost: sortedTariffsByCost,
        currentTariffWithCost,
        emailType: 'ALREADY_ON_CHEAPEST_TARIFF',
      });

      return logSuccessAndRespond(
        `You are already on the cheapest tariff: ${cheapestTariff.displayName} - ${todaysCostInPounds}`,
      );
    }

    const savingsInPence = todayCostInPence - cheapestTariff.costInPence;

    // Not worth switching for 2p
    if (savingsInPence <= 2) {
      await sendTariffEmail({
        allTariffsByCost: sortedTariffsByCost,
        currentTariffWithCost,
        emailType: 'NOT_WORTH_SWITCHING_TARIFF',
      });

      return logSuccessAndRespond(
        `Not worth switching to ${cheapestTariff.displayName} from ${currentTariff.displayName}`,
      );
    }

    if (!isDryRun()) {
      const enrolmentId = await getEnrollmentId({
        mpan,
        targetProductCode: cheapestTariff.productCode,
      });

      await sleep(60 * 1000);

      const acceptedVersion = await acceptNewAgreement({
        enrolmentId,
        productCode: cheapestTariff.productCode,
      });

      logger.info(`Accepted new tariff agreement: ${acceptedVersion}`);

      const isVerified = await verifyNewAgreement();

      if (!isVerified) {
        throw new AgreementVerificationError(
          'Unable to verify new agreement after multiple retries. Please check your account and emails.',
        );
      }
    }

    await sendTariffEmail({
      allTariffsByCost: sortedTariffsByCost,
      currentTariffWithCost,
      emailType: 'CHEAPER_TARIFF_EXISTS',
    });

    return logSuccessAndRespond(
      `Going to switch to ${cheapestTariff.displayName} - ${cheapestTariffCostInPounds} from ${currentTariff.displayName} - ${todaysCostInPounds}`,
    );
  } catch (error) {
    let message: string;

    if (error instanceof Error) {
      const err = error.toString();

      message = error.message;

      logger.error(err, error);
    } else {
      message = String(error);
    }

    return formatResponse(500, { message });
  }
}
