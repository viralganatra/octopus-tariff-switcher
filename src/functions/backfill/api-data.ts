import { isWithinInterval, parseISO } from 'date-fns';
import type { IsoDate } from '../../types/misc';
import type { ConsumptionIntervals, EletricityAgreements } from './schema';
import { chunkArray, processBatches } from '../../utils/batch';
import { TARIFFS, type TariffDisplayName } from '../../constants/tariff';
import { getProductByTariff } from '../tariff-switcher/api-data';
import { retryWithExponentialBackoff } from '../../utils/fetch';
import { fetchConsumption, fetchStandingCharge } from './queries';
import { fetchUnitRatesByTariff } from '../tariff-switcher/queries';
import type { TariffUnitRates } from '../tariff-switcher/schema';

type ItemCache = {
  isoDate: IsoDate;
  tariffCode: string;
  tariffName: TariffDisplayName;
  productCode: string;
  standingCharge: number;
  unitRates: TariffUnitRates;
  consumption: ConsumptionIntervals;
};

type ItemCacheBuilder = Pick<ItemCache, 'isoDate' | 'tariffCode' | 'tariffName'> &
  Partial<Omit<ItemCache, 'isoDate' | 'tariffCode' | 'tariffName'>>;

const BATCH_SIZE = 25;

export function findMatchingTariffForDate({
  pastTariffs,
  isoDate,
}: {
  pastTariffs: EletricityAgreements;
  isoDate: IsoDate;
}) {
  const targetDate = parseISO(isoDate);

  // Normalize target to start of day UTC
  const targetDateUtc = new Date(
    Date.UTC(targetDate.getUTCFullYear(), targetDate.getUTCMonth(), targetDate.getUTCDate()),
  );

  const matchingTariff = pastTariffs.find(({ validFrom, validTo }) => {
    const from = parseISO(validFrom);
    const to = parseISO(validTo);

    // isWithinInterval treats end as inclusive so we need to make it exclusive
    return (
      isWithinInterval(targetDateUtc, {
        start: from,
        end: to,
      }) && targetDateUtc.getTime() !== to.getTime()
    );
  });

  if (!matchingTariff) {
    throw new Error(`Unable to find matching tariff for date: ${isoDate}`);
  }

  return matchingTariff;
}

export async function enrichDatesWithTariffData({
  dates,
  pastTariffs,
  mpan,
  serialNumber,
}: { dates: IsoDate[]; pastTariffs: EletricityAgreements; mpan: string; serialNumber: string }) {
  const itemCacheBuilder = new Map<IsoDate, ItemCacheBuilder>();
  const batchesOfDates = chunkArray(dates, BATCH_SIZE);

  // Build initial cache directly
  for (const isoDate of dates) {
    const { tariffCode } = findMatchingTariffForDate({ pastTariffs, isoDate });

    const tariff = TARIFFS.find(({ tariffCodeMatcher }) => tariffCode.includes(tariffCodeMatcher));

    if (!tariff) {
      throw new Error(`No matching tariff for: ${tariffCode}`);
    }

    itemCacheBuilder.set(isoDate, {
      isoDate,
      tariffCode,
      tariffName: tariff.displayName,
    });
  }

  // Fetch product codes for unique tariff names
  const uniqueTariffNames = new Set<ItemCache['tariffName']>();

  for (const item of itemCacheBuilder.values()) {
    uniqueTariffNames.add(item.tariffName);
  }

  const tariffToProduct = new Map<ItemCache['tariffName'], string>();

  await Promise.all(
    [...uniqueTariffNames].map(async (tariffName) => {
      const { code } = await getProductByTariff(tariffName);

      tariffToProduct.set(tariffName, code);
    }),
  );

  // Add product codes to cache
  for (const [isoDate, item] of itemCacheBuilder) {
    const productCode = tariffToProduct.get(item.tariffName);

    itemCacheBuilder.set(isoDate, { ...item, productCode });
  }

  // Fetch standing charge, unit rates & consumption in parallel batches
  const unitAndStandingChargesForDates = await processBatches({
    batches: batchesOfDates,
    processBatchItem: (isoDate) =>
      retryWithExponentialBackoff(async () => {
        const item = itemCacheBuilder.get(isoDate);

        if (!item || !item.productCode) {
          throw new Error(`Missing item or productCode for date: ${isoDate}`);
        }

        const { tariffCode, productCode } = item;

        const [standingCharge, unitRates, consumption] = await Promise.all([
          fetchStandingCharge({ isoDate, tariffCode, productCode }),
          fetchUnitRatesByTariff({ isoDate, tariffCode, productCode }),
          fetchConsumption({ isoDate, mpan, serialNumber }),
        ]);

        return { isoDate, standingCharge, unitRates, consumption };
      }),
  });

  // Populate item cache with standing charge, unit rates & consumption
  const itemCache = new Map<IsoDate, ItemCache>();

  for (const {
    isoDate,
    standingCharge,
    unitRates,
    consumption,
  } of unitAndStandingChargesForDates) {
    const item = itemCacheBuilder.get(isoDate);

    if (!item || !item.productCode) {
      throw new Error(`Missing cache item for date: ${isoDate}`);
    }

    itemCache.set(isoDate, {
      isoDate,
      standingCharge,
      unitRates,
      consumption,
      tariffCode: item.tariffCode,
      tariffName: item.tariffName,
      productCode: item.productCode,
    });
  }

  return itemCache;
}
