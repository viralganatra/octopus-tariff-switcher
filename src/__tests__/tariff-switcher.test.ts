import { graphql, http, HttpResponse } from 'msw';
import type { APIGatewayProxyEvent, Context } from 'aws-lambda';
import { server } from '../mocks/node';
import { tariffSwitcher } from '../tariff-switcher';
import * as email from '../notifications/email';
import * as ApiData from '../functions/tariff-switcher/api-data';
import { accountFixture, productGoFixture } from '../mocks/fixtures';

describe('Tariff Switcher', () => {
  const proxy = {} as APIGatewayProxyEvent;
  const context = {} as Context;

  beforeEach(() => {
    vi.setSystemTime(new Date(2025, 2, 3));
  });

  it('should switch to the cheapest tariff and send a notification', async () => {
    const spy = vi.spyOn(email, 'sendEmail');
    const promise = tariffSwitcher(proxy, context);

    await vi.runAllTimersAsync();

    const data = await promise;

    // @ts-ignore
    const emailData = spy.mock.calls[0][0];

    expect(emailData).toMatchSnapshot();
    expect(data.body).toMatchInlineSnapshot(`
      "{
        "message": "Going to switch to Octopus Go - £0.47 from Agile Octopus - £0.80"
      }"
    `);
  });

  it('should throw an error if the new agreement cannot be verified', async () => {
    const fixture = structuredClone(accountFixture);

    // @ts-ignore
    fixture.account.electricityAgreements[0].validFrom = '2025-03-08T00:00:00+00:00';

    server.use(
      graphql.query('Account', () => {
        return HttpResponse.json({
          data: fixture,
        });
      }),
    );

    const promise = tariffSwitcher(proxy, context);

    await vi.runAllTimersAsync();

    const data = await promise;

    expect(data).toMatchInlineSnapshot(`
      {
        "body": "{
        "message": "Unable to verify new agreement after multiple retries. Please check your account and emails."
      }",
        "statusCode": 500,
      }
    `);
  });

  it('should not switch the tariff if it is already the cheapest and send a notification', async () => {
    const spy = vi.spyOn(email, 'sendEmail');
    const fixture = structuredClone(productGoFixture);

    fixture.single_register_electricity_tariffs._A.direct_debit_monthly.standing_charge_inc_vat = 48.788145;

    server.use(
      http.get('https://api.octopus.energy/v1/products/GO-VAR-22-10-14/', () => {
        return HttpResponse.json(fixture);
      }),
    );

    const promise = tariffSwitcher(proxy, context);

    await vi.runAllTimersAsync();

    const data = await promise;

    // @ts-ignore
    const emailData = spy.mock.calls[0][0];

    expect(emailData).toMatchSnapshot();
    expect(data.body).toMatchInlineSnapshot(`
      "{
        "message": "You are already on the cheapest tariff: Agile Octopus - £0.80"
      }"
    `);
  });

  it('should not switch the tariff if the difference is less than 2p and send a notification', async () => {
    const spy = vi.spyOn(email, 'sendEmail');
    const fixture = structuredClone(productGoFixture);

    fixture.single_register_electricity_tariffs._A.direct_debit_monthly.standing_charge_inc_vat = 40;

    server.use(
      http.get('https://api.octopus.energy/v1/products/GO-VAR-22-10-14/', () => {
        return HttpResponse.json(fixture);
      }),
    );

    const promise = tariffSwitcher(proxy, context);

    await vi.runAllTimersAsync();

    const data = await promise;

    // @ts-ignore
    const emailData = spy.mock.calls[0][0];

    expect(emailData).toMatchSnapshot();
    expect(data.body).toMatchInlineSnapshot(`
      "{
        "message": "Not worth switching to Octopus Go from Agile Octopus"
      }"
    `);
  });

  it('should catch all errors', async () => {
    server.use(
      http.get('https://api.octopus.energy/v1/products/COSY-22-12-08/', () => HttpResponse.error()),
    );

    const promise = tariffSwitcher(proxy, context);

    await vi.runAllTimersAsync();

    const data = await promise;

    expect(data).toMatchInlineSnapshot(`
      {
        "body": "{
        "message": "Failed to fetch"
      }",
        "statusCode": 500,
      }
    `);
  });

  describe('DRY_RUN', () => {
    beforeAll(() => {
      vi.stubEnv('DRY_RUN', 'true');
    });

    afterAll(() => {
      vi.stubEnv('DRY_RUN', 'false');
    });

    it('should not send an email if the DRY_RUN param is true', async () => {
      const spy = vi.spyOn(email, 'sendEmail');
      const promise = tariffSwitcher(proxy, context);

      await vi.runAllTimersAsync();
      await promise;

      expect(spy).not.toHaveBeenCalled();
    });

    it('should not switch the tariff if the DRY_RUN param is true', async () => {
      const spy = vi.spyOn(ApiData, 'getEnrollmentId');
      const promise = tariffSwitcher(proxy, context);

      await vi.runAllTimersAsync();
      await promise;

      expect(spy).not.toHaveBeenCalled();
    });
  });
});
