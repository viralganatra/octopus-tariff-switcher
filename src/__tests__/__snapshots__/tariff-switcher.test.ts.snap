// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tariff Switcher > should not switch the tariff if it is already the cheapest and send a notification 1`] = `
{
  "allTariffsByCost": [
    {
      "cost": 80.324,
      "displayName": "Agile Octopus",
      "id": "agile",
      "productCode": "AGILE-24-10-01",
      "tariffCodeMatcher": "-AGILE-",
    },
    {
      "cost": 87.452,
      "displayName": "Cosy Octopus",
      "id": "cosy",
      "productCode": "COSY-22-12-08",
      "tariffCodeMatcher": "-COSY-",
    },
    {
      "cost": 87.452,
      "displayName": "Octopus Go",
      "id": "go",
      "productCode": "GO-VAR-22-10-14",
      "tariffCodeMatcher": "-GO-",
    },
  ],
  "currentTariffWithCost": {
    "cost": 80.324,
    "displayName": "Agile Octopus",
    "id": "agile",
    "productCode": "AGILE-24-10-01",
    "tariffCodeMatcher": "-AGILE-",
  },
  "emailType": "ALREADY_ON_CHEAPEST_TARIFF",
}
`;

exports[`Tariff Switcher > should not switch the tariff if the difference is less than 2p and send a notification 1`] = `
{
  "allTariffsByCost": [
    {
      "cost": 78.6639,
      "displayName": "Octopus Go",
      "id": "go",
      "productCode": "GO-VAR-22-10-14",
      "tariffCodeMatcher": "-GO-",
    },
    {
      "cost": 80.324,
      "displayName": "Agile Octopus",
      "id": "agile",
      "productCode": "AGILE-24-10-01",
      "tariffCodeMatcher": "-AGILE-",
    },
    {
      "cost": 87.452,
      "displayName": "Cosy Octopus",
      "id": "cosy",
      "productCode": "COSY-22-12-08",
      "tariffCodeMatcher": "-COSY-",
    },
  ],
  "currentTariffWithCost": {
    "cost": 80.324,
    "displayName": "Agile Octopus",
    "id": "agile",
    "productCode": "AGILE-24-10-01",
    "tariffCodeMatcher": "-AGILE-",
  },
  "emailType": "NOT_WORTH_SWITCHING_TARIFF",
}
`;

exports[`Tariff Switcher > should switch to the cheapest tariff and send a notification 1`] = `
{
  "allTariffsByCost": [
    {
      "cost": 47.452,
      "displayName": "Octopus Go",
      "id": "go",
      "productCode": "GO-VAR-22-10-14",
      "tariffCodeMatcher": "-GO-",
    },
    {
      "cost": 80.324,
      "displayName": "Agile Octopus",
      "id": "agile",
      "productCode": "AGILE-24-10-01",
      "tariffCodeMatcher": "-AGILE-",
    },
    {
      "cost": 87.452,
      "displayName": "Cosy Octopus",
      "id": "cosy",
      "productCode": "COSY-22-12-08",
      "tariffCodeMatcher": "-COSY-",
    },
  ],
  "currentTariffWithCost": {
    "cost": 80.324,
    "displayName": "Agile Octopus",
    "id": "agile",
    "productCode": "AGILE-24-10-01",
    "tariffCodeMatcher": "-AGILE-",
  },
  "emailType": "CHEAPER_TARIFF_EXISTS",
}
`;
